freud/voronoi.pyx

```diff
@@ -322,41 +313,56 @@ class Voronoi(object):
-            if added_j:
-                self.firstShellWeight[index_j].append(weight)
+            all_bonds[(index_i, index_j)] = max(all_bonds.get(
+                                                (index_i, index_j), 0), weight)
+            all_bonds[(index_j, index_i)] = max(all_bonds.get(
+                                                (index_j, index_i), 0), weight)
+
+        # Build neighbor list based on voronoi neighbors
+        cdef np.ndarray[np.float32_t, ndim=1] weights = \
+            np.asarray(list(all_bonds.values()), dtype=np.float32)
+        cdef np.ndarray[np.uint64_t, ndim=2] bond_indices = \
+            np.asarray(list(all_bonds.keys()), dtype=np.uint64)
 
+        cdef np.ndarray[np.long_t, ndim=1] sort_indices = \
+            np.lexsort((bond_indices[:, 1], bond_indices[:, 0]))
+
+        bond_indices = bond_indices[sort_indices]
+        weights = weights[sort_indices]
+
+        self._nlist = freud.locality.NeighborList.from_arrays(
+            np.max(bond_indices[:, 0])+1, np.max(bond_indices[:, 1])+1,
+            bond_indices[:, 0], bond_indices[:, 1], weights=weights)
         return self
 
     def getNeighbors(self, numShells):
-        R"""Get :code:`numShells` of neighbors for each particle
+        R"""Get well-sorted neighbors from cumulative Voronoi shells for each
+        particle by specifying :code:`numShells`.
 
         Must call :meth:`~.computeNeighbors()` before this method.
 
         Args:
             numShells (int): Number of neighbor shells.
         """
-        neighbor_list = copy.copy(self.firstShellNeighborList)
-        # delete [] in neighbor_list
-        neighbor_list = [x for x in neighbor_list if len(x) > 0]
-        for _ in range(numShells - 1):
-            dummy_neighbor_list = copy.copy(neighbor_list)
-            for i in range(len(neighbor_list)):
-                numNeighbors = len(neighbor_list[i])
-                for j in range(numNeighbors):
-                    dummy_neighbor_list[i] = dummy_neighbor_list[i] + \
-                        self.firstShellNeighborList[neighbor_list[i][j]]
-
-                # remove duplicates
-                dummy_neighbor_list[i] = list(set(dummy_neighbor_list[i]))
-
-                if i in dummy_neighbor_list[i]:
-                    dummy_neighbor_list[i].remove(i)
-
-            neighbor_list = copy.copy(dummy_neighbor_list)
-
-        return neighbor_list
+        nlist = self.nlist
+        max_index = np.max(nlist[:]+1)
+        # convert neighbor list to a csr matrix and set weights = 1
+        sparse_neighbors = csr_matrix(
+            (np.ones(len(nlist)), (nlist.index_i, nlist.index_j)),
+            shape=(max_index, max_index))
+        # take a numShell power of the matrix
+        # sum over all shell neighbors, and convert to a linked list
+        sparse_neighbors = sum(
+            [sparse_neighbors**k for k in range(1, numShells+1)])
+        lil_neighbors = sparse_neighbors.tolil(copy=False)
```

Comment:
Can you remove the `copy=False`? I think we actually want a copy in most cases, since we want to convert it to the list-of-lists arrangement in memory. Then we can index `lil_neighbors` efficiently.

Reply:
Also the tests failed on Python 3.5, perhaps because older versions of SciPy don't have the `copy` argument. Another reason to drop it.